packages/ragbits-document-search/src/ragbits/document_search/ingestion/providers/unstructured/default.py

```diff
@@ -132,20 +137,26 @@ async def process(self, document_meta: DocumentMeta) -> list[Element]:
                 **self.partition_kwargs,
             )
 
-        elements = chunk_elements(elements, **self.chunking_kwargs)
-        return [_to_text_element(element, document_meta) for element in elements]
-
-
-def _to_text_element(element: UnstructuredElement, document_meta: DocumentMeta) -> TextElement:
-    return TextElement(
-        document_meta=document_meta,
-        content=element.text,
-    )
-
-
-def _set_or_raise(name: str, value: Optional[str], env_var: str) -> str:
-    if value is not None:
-        return value
-    if (env_value := os.getenv(env_var)) is None:
-        raise ValueError(f"Either pass {name} argument or set the {env_var} environment variable")
-    return env_value
+        return await self._chunk_and_convert(elements, document_meta, document.local_path)
+
+    async def _chunk_and_convert(
+        self, elements: list[UnstructuredElement], document_meta: DocumentMeta, document_path: Path
+    ) -> list[Element]:
+        if self.__class__ == UnstructuredDefaultProvider:
```

Comment:
This looks confusing to me. Is this about having a behavior on the base class that the classes inheriting from it don't share? 

Inheriting a class and having the behavior change magically, even though I haven't changed it myself, would be surprising and not something I would like to encounter as a programmer. Seems like this behavior should be in a method that inheriting class can overwrite if they want.

Reply:
Both inheriting classes need to do the same preprocessing:
```
        image_elements = [e for e in elements if e.category == ElementType.IMAGE]
        other_elements = [e for e in elements if e.category != ElementType.IMAGE]
        chunked_other_elements = chunk_elements(other_elements, **self.chunking_kwargs)

        text_elements: list[Element] = [to_text_element(element, document_meta) for element in chunked_other_elements]
```
I didn't want to duplicate this code in both classes nor introduce a third "in-between" class so I came up with this idea. Happy to change it if you have a better suggestion how to handle it.

Reply:
I would definitely change this - it's a very surprising behavior and you don't want unexpected surprises from your code. Even though both classes that **currently** inherit from `UnstructuredDefaultProvider` need this behavior, the base class should be ready to be subclassed any time for other reasons. It shouldn't contain code specific to functionality that is not related to its intended behavior and definitely shouldn't magically switch to it for all of its children. 

This is an important principle: child classes are downstream from the base class and they can know and use the specific of the base class. On the other hand, the base class is upstream from its children. It cannot know when it will be subclassed and for what reason. 

Seems that `UnstructuredPdfProvider` and `UnstructuredImageProvider` are similar because they both deal with images - the only difference is where they get the images from.

You can reflect that in the inheritance structure. Maybe pdf provider is just an image provider enhanced with the ability to find images in PDFs? Than the inheritance could look like that:

```
UnstructuredDefaultProvider -> UnstructuredImageProvider -> UnstructuredPdfProvider
``` 

If this doesn't seem correct to you in this case, alternatively you could have an intermediate (possibly abstract) class that adds image-related functionality, inherits from `UnstructuredDefaultProvider` and is a base for both `UnstructuredImageProvider` and `UnstructuredPdfProvider`

Reply:
Thanks, that turned out to be a good idea, now pdf extends the images and there are only two small functions that needed to be implemented